lib/RAST_SDK/RAST_SDKImpl.pm

```diff
@@ -1454,143 +1454,145 @@ sub annotate_genomes
-					ref=>Bio::KBase::kbaseenv::buildref($params->{workspace},$ref)}])->[0]->{'refs'};
-
-				if (ref($obj) eq 'HASH') {
-					foreach my $key (keys %$obj) {
-						if ( grep( /^$key$/, @$replace_genomes ) ) {
-							$warn .= "WARNING: Found Duplicate Genome $key";
-						} else {
-							push(@$replace_genomes,$key);
-						}
-					}
-				} elsif (ref($obj) eq 'ARRAY') {
-					foreach my $key (@$obj) {
-						if ( grep( /^$key$/, @$replace_genomes ) ) {
-							$warn .= "WARNING: Found Duplicate Assembly $key";
-						} else {
-							push(@$replace_genomes,$key);
-						}
-					}
-					
-				}
-			} else {
-				if ( grep( /^$ref$/, @$replace_genomes ) ) {
-					$warn .= "WARNING: Found Duplicate Assembly $ref";
-				} else {
-					push(@$replace_genomes,$ref);
-				}
-			}
-		}
-		print STDERR "WARNiNG $warn\n";
-		$genomes = $replace_genomes;
-	}
+    if ($params->{ncbi_taxon_id}) {
+        $params->{scientific_name} = $self->get_scientific_name_for_NCBI_taxon(
+            $params->{ncbi_taxon_id}, $params->{relation_engine_timestamp_ms});
+    }
+    my $htmlmessage = "";
+    my $warn        = "";
+    my $genomes     = $params->{input_genomes};
+    my $obj_type;
+
+    #
+    # If $genomes is an ARRAY, then multiple genomes or assemblies or sets were submitted
+    #
+    # When a set is submitted, the set needs to be unpacked and added to the list
+    # Create a non-redundant replacement list:
+    #	1. Individual genomes and assemblies are added
+    #	2. Iterate over a Set to add to the replacement list
+    #	3. GenomeSets use a HASH and AssemblySets use an ARRAY
+    #	4. Use perl grep to see if the ref is already in the list
+    #	5. Issue a warning when a duplicate is found so user knows what happened.
+    #
+    my $empty_input_msg = "Genomes expected in an array with at least one genome as input objects,";
+    $empty_input_msg .= " and/or in a semicolon (;) delimited string of genome names/ids.";
+    Bio::KBase::Exceptions::ArgumentValidationError->throw(
+	error        => $empty_input_msg,
+	method_name  => 'annotate_genomes'
+    ) unless (is_arrayref( $genomes ) && @$genomes) || defined($params->{genome_text});
+
+    my $replace_genomes = [];
+    foreach my $ref (@$genomes) {
+        my $info = Bio::KBase::kbaseenv::get_object_info([{ref=>$ref}],0);
+        $obj_type = $info->[0]->[2];
+	if ($obj_type =~ /KBaseSearch\.GenomeSet/ || $obj_type =~ /KBaseSets\.AssemblySet/) {
+            my $obj = Bio::KBase::kbaseenv::get_objects([{
+                ref=>Bio::KBase::kbaseenv::buildref($params->{workspace},$ref)}])->[0]->{'refs'};
+
+            if (ref($obj) eq 'HASH') {
+                foreach my $key (keys %$obj) {
+                    if ( grep( /^$key$/, @$replace_genomes ) ) {
+                        $warn .= "WARNING: Found Duplicate Genome $key";
+                    } else {
+                        push(@$replace_genomes,$key);
+                    }
+                }
```

Comment:
(Pre-existing code)
Hash keys will always be unique, so this could be replaced by
```
@$replace_genomes = keys %$obj;
```